### PR TITLE
[ARM CPU] Add ACL executor for Convert

### PR DESCRIPTION
**Pull request by allnes**
- Separate executors
- Add ACL executor for convert

### CODE REVIEW

**Comment by akladiev** [comment]
This PR will be closed in a week because of 2 weeks of no activity.

File: src/plugins/intel_cpu/src/nodes/convert.cpp [hunk 1]
```diff
@@ -144,15 +152,27 @@ void Convert::execute(dnnl::stream strm) {
...
-                dstPtr,
-                parentMem.getDesc().getPrecision(),
-                origPrc,
-                childMem.getDesc().getPrecision(),
-                parentPaddElemCount);
+    dnnl::primitive_attr attr;
+    auto selectedPD = getSelectedPrimitiveDescriptor();
+    std::vector<MemoryDescPtr> srcDescs = {parentMem.getDescPtr()};
+    std::vector<MemoryDescPtr> dstDescs = {childMem.getDescPtr()};
+    execPtr = selectedPD->getExecutorFactoryAs<ConvertExecutorFactory>()->makeExecutor(convertParams,
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/nodes/convert.cpp:170]: Could you please create a prepareParams method for the convert node to align with the rest of the nodes.

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class ACLConvertExecutorBuilder : public ConvertExecutorBuilder {
+public:
+    ~ACLConvertExecutorBuilder() = default;
+    bool isSupported(const ConvertParams& convertParams,
+                     const std::vector<MemoryDescPtr>& srcDescs,
+                     const std::vector<MemoryDescPtr>& dstDescs) const override {
+        if (convertParams.srcPrc != convertParams.dstPrc) {
+            if (convertParams.srcPrc != InferenceEngine::Precision::I8 &&
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp:41]: Please use one_of() utils method for such cases
Here and below

**Review by EgorDuplensky — COMMENTED** [review]
Could you please execute dlb regression (with timeout = 1) just to verify functionality.

File: src/plugins/intel_cpu/src/nodes/convert.cpp (hunk 1 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/convert.cpp]: fixed

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp (hunk 2 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp]: fixed

File: src/plugins/intel_cpu/src/utils/cpu_utils.hpp [hunk 3]
```diff
@@ -122,7 +122,7 @@ inline InferenceEngine::Precision normalizeToSupportedPrecision(InferenceEngine:
             break;
         }
         case InferenceEngine::Precision::FP16: {
-            precision = InferenceEngine::Precision::FP32;
+            //precision = InferenceEngine::Precision::FP32;
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/utils/cpu_utils.hpp:125]: please revert

**Review by EgorDuplensky — APPROVED** [review]

File: src/plugins/intel_cpu/src/transformations/transformation_pipeline.cpp [hunk 4]
```diff
@@ -217,7 +217,7 @@ void Transformations::PreLpt(const std::vector<ov::element::Type>& defaultPrecis
             {ov::element::u16,     ov::element::i32},
             {ov::element::u32,     ov::element::i32},
             {ov::element::f64,     ov::element::f32},
-            {ov::element::f16,     ov::element::f32},
+            //{ov::element::f16,     ov::element::f32},
```
[the remark this answers is not in the document]
Comment (alvoron) [reply src/plugins/intel_cpu/src/transformations/transformation_pipeline.cpp]: done

File: src/plugins/intel_cpu/src/nodes/convert.cpp [hunk 5]
```diff
@@ -144,15 +168,16 @@ void Convert::execute(dnnl::stream strm) {
...
-    void* srcPtr = parentMem.GetPtr();
-    void* dstPtr = childMem.GetPtr();
+    std::vector<MemoryCPtr> srcMemory;
+    for (size_t i = 0; i < getOriginalInputsNumber(); i++) {
+        srcMemory.push_back(getParentEdgeAt(i)->getMemoryPtr());
+    }
+    std::vector<MemoryPtr> dstMemory;
+    for (size_t i = 0; i < getOriginalOutputsNumber(); i++) {
+        dstMemory.push_back(getChildEdgeAt(i)->getMemoryPtr());
+    }
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/convert.cpp:178]: In prepare params we use only zero port, but here all the ports. May be it does make sense to do it uniformly?

File: src/plugins/intel_cpu/src/nodes/convert.h [hunk 6]
```diff
@@ -40,7 +42,7 @@ class Convert : public Node {
     const MemoryDesc& getInput() const { return *input; }
     const MemoryDesc& getOutput() const { return *output; }
 
-    bool needPrepareParams() const override { return false; }
+    bool needPrepareParams() const override { return true; }
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/convert.h:45]: Do we really need to do it unconditionally even if the shape hasn't changed?

File: src/plugins/intel_cpu/src/nodes/executors/convert_list.hpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                             const ExecutorContext::CPtr context) : ExecutorFactory(context) {
+        for (auto& desc : getConvertExecutorsList()) {
+            if (desc.builder->isSupported(convertParams, srcDescs, dstDescs)) {
+                supportedDescs.push_back(desc);
+            }
+        }
+    }
+
+    ~ConvertExecutorFactory() = default;
+    virtual ConvertExecutorPtr makeExecutor(const ConvertParams& convertParams,
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/convert_list.hpp:43]: It doesn't make much sense to keep this method implementation in header.

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    bool init(const ConvertParams& convertParams,
+              const std::vector<MemoryDescPtr>& srcDescs,
+              const std::vector<MemoryDescPtr>& dstDescs,
+              const dnnl::primitive_attr &attr) override;
+    void exec(const std::vector<MemoryCPtr>& src, const std::vector<MemoryPtr>& dst) override;
+    impl_desc_type getImplType() const override { return implDescType; };
+    ~CommonConvertExecutor() override = default;
+protected:
+    ConvertParams commonConvertParams;
+    impl_desc_type implDescType = impl_desc_type::unknown;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp:24]: Why `impl_desc_type::unknown;` is used while this is a reference implementation? AFAIK we do have `ref` implementation type.

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+#pragma once
+
+#include "nodes/executors/convert.hpp"
+
+namespace ov {
+namespace intel_cpu {
+
+class CommonConvertExecutor : public ConvertExecutor {
+public:
+    explicit CommonConvertExecutor(const ExecutorContext::CPtr context);
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp:14]: This constructor may be inherited from the base class.

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    bool init(const ConvertParams& convertParams,
+              const std::vector<MemoryDescPtr>& srcDescs,
+              const std::vector<MemoryDescPtr>& dstDescs,
+              const dnnl::primitive_attr &attr) override;
+    void exec(const std::vector<MemoryCPtr>& src, const std::vector<MemoryPtr>& dst) override;
+    impl_desc_type getImplType() const override { return implDescType; };
+    ~CommonConvertExecutor() override = default;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp:21]: This declaration doesn't make much sense.

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    arm_compute::Tensor srcTensor, dstTensor;
+    std::unique_ptr<arm_compute::NECopy> acl_copy;
+    std::unique_ptr<arm_compute::NECast> acl_cast;
+};
+
+
+class ACLConvertExecutorBuilder : public ConvertExecutorBuilder {
+public:
+    ~ACLConvertExecutorBuilder() = default;
+    bool isSupported(const ConvertParams& convertParams,
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp:37]: Do we really need to keep this method definition in the header?

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+#include "nodes/executors/convert.hpp"
+#include "utils/debug_capabilities.h"
+#include "arm_compute/runtime/NEON/NEFunctions.h"
+
+namespace ov {
+namespace intel_cpu {
+
+class ACLConvertExecutor : public ConvertExecutor {
+public:
+    explicit ACLConvertExecutor(const ExecutorContext::CPtr context);
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp:16]: The constructor may be inherited

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    bool init(const ConvertParams& convertParams,
+              const std::vector<MemoryDescPtr>& srcDescs,
+              const std::vector<MemoryDescPtr>& dstDescs,
+              const dnnl::primitive_attr &attr) override;
+    void exec(const std::vector<MemoryCPtr>& src, const std::vector<MemoryPtr>& dst) override;
+    impl_desc_type getImplType() const override { return implDescType; };
+    ~ACLConvertExecutor() override = default;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp:23]: This declaration doesn't make much sense.

File: src/plugins/intel_cpu/src/nodes/convert.cpp (hunk 5 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/convert.cpp:182]: fixed

File: src/plugins/intel_cpu/src/nodes/convert.h (hunk 6 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/convert.h]: fixed

File: src/plugins/intel_cpu/src/nodes/executors/convert_list.hpp (hunk 7 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/convert_list.hpp:43]: moved

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp (hunk 8 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp]: changed to ref

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp (hunk 9 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp]: fixed

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp (hunk 10 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp]: removed

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp (hunk 12 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp]: fixed

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp (hunk 13 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp]: removed

File: src/plugins/intel_cpu/src/nodes/convert.cpp [hunk 14]
```diff
@@ -123,13 +131,33 @@ void Convert::initSupportedPrimitiveDescriptors() {
...
+
+    auto selectedPD = getSelectedPrimitiveDescriptor();
+    std::vector<MemoryDescPtr> srcDescs;
+    for (size_t i = 0; i < getOriginalInputsNumber(); i++) {
+        srcDescs.push_back(getParentEdgeAt(i)->getMemoryPtr()->getDescPtr());
+    }
+    std::vector<MemoryDescPtr> dstDescs;
+    for (size_t i = 0; i < getOriginalOutputsNumber(); i++) {
+        dstDescs.push_back(getChildEdgeAt(i)->getMemoryPtr()->getDescPtr());
+    }
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/convert.cpp:153]: `OriginalInputsNumber` and `riginalOutputsNumber` doesn't necessarily correspond to the actual number of input and output ports of the CPU node. It's more safe to use port configs. However, given the semantics of the `Convert` operation, does it have more than one input and one output ports? If so, does it make sense to use vectors then?

File: src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    using ConvertExecutor::ConvertExecutor;
+    bool init(const ConvertParams& convertParams,
+              const std::vector<MemoryDescPtr>& srcDescs,
+              const std::vector<MemoryDescPtr>& dstDescs,
+              const dnnl::primitive_attr &attr) override;
+    void exec(const std::vector<MemoryCPtr>& src, const std::vector<MemoryPtr>& dst) override;
+    impl_desc_type getImplType() const override { return implDescType; };
+protected:
+    ConvertParams commonConvertParams;
+    impl_desc_type implDescType = impl_desc_type::ref;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/common/ref_convert.hpp:23]: `static const` ?

File: src/plugins/intel_cpu/src/nodes/executors/convert.hpp [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+protected:
+    ConvertParams convertParams;
+    const ExecutorContext::CPtr convertContext;
+};
+using ConvertExecutorPtr = std::shared_ptr<ConvertExecutor>;
+using ConvertExecutorCPtr = std::shared_ptr<const ConvertExecutor>;
+
+class ConvertExecutorBuilder {
+public:
+    ~ConvertExecutorBuilder() = default;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/convert.hpp:40]:
```suggestion
    virtual ~ConvertExecutorBuilder() = default;
```

File: src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    bool init(const ConvertParams& convertParams,
+              const std::vector<MemoryDescPtr>& srcDescs,
+              const std::vector<MemoryDescPtr>& dstDescs,
+              const dnnl::primitive_attr &attr) override;
+    void exec(const std::vector<MemoryCPtr>& src, const std::vector<MemoryPtr>& dst) override;
+    impl_desc_type getImplType() const override { return implDescType; };
+protected:
+    ConvertParams aclConvertParams;
+    bool isCopyOp;
+    impl_desc_type implDescType = impl_desc_type::acl;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/executors/acl/acl_convert.hpp:26]: `static const`?

File: src/plugins/intel_cpu/src/nodes/executors/convert.hpp (hunk 16 above)
Comment (alvoron) [reply src/plugins/intel_cpu/src/nodes/executors/convert.hpp:40]: fixed

**Pull request merged**